core/src/main/java/org/infinispan/notifications/cachelistener/DistributedQueueingSegmentListener.java

```diff
@@ -137,8 +115,34 @@ public boolean handleEvent(EventWrapper<K, V, CacheEntryEvent<K, V>> wrapped, Li
    }
 
    @Override
-   public void accept(Supplier<PrimitiveIterator.OfInt> segments) {
-      justCompletedSegments = segments.get();
+   public void accept(int segment) {
+      if (justCompletedSegments == null) {
+         justCompletedSegments = IntSets.mutableEmptySet();
+      }
+      justCompletedSegments.set(segment);
+   }
+
+   @Override
+   public CompletionStage<Void> delayProcessing() {
+      AggregateCompletionStage<Void> aggregateCompletionStage = null;
+      if (justCompletedSegments != null) {
+         if (trace) {
+            log.tracef("Segments %s completed for listener", justCompletedSegments);
+         }
+         // This relies on the fact that notifiedKey is immediately called after the entry has finished being iterated on
+         PrimitiveIterator.OfInt iter = justCompletedSegments.iterator();
+         while (iter.hasNext()) {
+            CompletionStage<Void> segmentStage = completeSegment(iter.nextInt());
+            if (segmentStage != null) {
+               if (aggregateCompletionStage == null) {
```

Comment:
I won't revert it now, but if this stage is not null there is a high chance that it will be complete as most of our listeners in this case will be complete.